pkg/ruler/ruler.go

```diff
@@ -116,92 +120,43 @@ func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
-	alertURL      *url.URL
-	notifierCfg   *config.Config
-	queueCapacity int
-	groupTimeout  time.Duration
-	metrics       *rules.Metrics
+	cfg              Config
+	engine           *promql.Engine
+	queryable        storage.Queryable
+	pusher           Pusher
+	alertURL         *url.URL
+	notifierCfg      *config.Config
+	queueCapacity    int
+	groupTimeout     time.Duration
+	SearchPendingFor time.Duration
+
+	scheduler *scheduler
+	workers   []worker
+
+	lifecycler *ring.Lifecycler
+	ring       *ring.Ring
 
 	// Per-user notifiers with separate queues.
 	notifiersMtx sync.Mutex
 	notifiers    map[string]*rulerNotifier
 }
 
-// rulerNotifier bundles a notifier.Manager together with an associated
-// Alertmanager service discovery manager and handles the lifecycle
-// of both actors.
-type rulerNotifier struct {
-	notifier  *notifier.Manager
-	sdCancel  context.CancelFunc
-	sdManager *discovery.Manager
-	wg        sync.WaitGroup
-	logger    gklog.Logger
-}
-
-func newRulerNotifier(o *notifier.Options, l gklog.Logger) *rulerNotifier {
-	sdCtx, sdCancel := context.WithCancel(context.Background())
-	return &rulerNotifier{
-		notifier:  notifier.NewManager(o, l),
-		sdCancel:  sdCancel,
-		sdManager: discovery.NewManager(sdCtx, l),
-		logger:    l,
-	}
-}
-
-func (rn *rulerNotifier) run() {
-	rn.wg.Add(2)
-	go func() {
-		if err := rn.sdManager.Run(); err != nil {
-			level.Error(rn.logger).Log("msg", "error starting notifier discovery manager", "err", err)
-		}
-		rn.wg.Done()
-	}()
-	go func() {
-		rn.notifier.Run(rn.sdManager.SyncCh())
-		rn.wg.Done()
-	}()
-}
-
-func (rn *rulerNotifier) applyConfig(cfg *config.Config) error {
-	if err := rn.notifier.ApplyConfig(cfg); err != nil {
-		return err
-	}
-
-	sdCfgs := make(map[string]sd_config.ServiceDiscoveryConfig)
-	for _, v := range cfg.AlertingConfig.AlertmanagerConfigs {
-		// AlertmanagerConfigs doesn't hold an unique identifier so we use the config hash as the identifier.
-		b, err := json.Marshal(v)
-		if err != nil {
-			return err
-		}
-		// This hash needs to be identical to the one computed in the notifier in
-		// https://github.com/prometheus/prometheus/blob/719c579f7b917b384c3d629752dea026513317dc/notifier/notifier.go#L265
-		// This kind of sucks, but it's done in Prometheus in main.go in the same way.
-		sdCfgs[fmt.Sprintf("%x", md5.Sum(b))] = v.ServiceDiscoveryConfig
-	}
-	return rn.sdManager.ApplyConfig(sdCfgs)
-}
-
-func (rn *rulerNotifier) stop() {
-	rn.sdCancel()
-	rn.notifier.Stop()
-	rn.wg.Wait()
-}
-
 // NewRuler creates a new ruler from a distributor and chunk store.
-func NewRuler(cfg Config, engine *promql.Engine, queryable storage.Queryable, d *distributor.Distributor) (*Ruler, error) {
+func NewRuler(cfg Config, engine *promql.Engine, queryable storage.Queryable, d *distributor.Distributor, rulesAPI RuleStore) (*Ruler, error) {
 	ncfg, err := buildNotifierConfig(&cfg)
 	if err != nil {
 		return nil, err
 	}
-	return &Ruler{
+
+	ruler := &Ruler{
+		cfg:           cfg,
```

Comment:
Now you're embedding the config (good), can you remove the fields that are just copied out of config?